### PR TITLE
Check quotas before creating platforms

### PR DESCRIPTION
**Pull request by mkjpryor**
This PR implements quota checks for CaaS cluster creation only, but with an aim to be easily extensible to Kubernetes + updates.

### CODE REVIEW

File: api/azimuth/scheduling/dto.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    #: The number of machines consumed by the platform
+    machines: int
+    #: The number of volumes consumed by the platform
+    volumes: int
+    #: The number of external IPs consumed by the platform
+    external_ips: int
+    #: The number of CPUs consumed by the platform
+    cpus: int
+    #: The amount of RAM (in MB) consumed by the platform
+    ram: int
```
Comment (JohnGarbutt) [inline api/azimuth/scheduling/dto.py:44]: Can we call this ram_mb and storage_gb to avoid mistakes?

File: api/azimuth/scheduling/dto.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    #: The amount of storage (in GB) consumed by the platform
+    storage: int
```
Comment (JohnGarbutt) [inline api/azimuth/scheduling/dto.py:46]: ... but there are many different types of storage, what does this mean? just volumes? or disk_gb in Nova too?

File: api/azimuth/scheduling/dto.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    Represents a summary of resource.
+    """
+    #: The number of machines consumed by the platform
+    machines: int
+    #: The number of volumes consumed by the platform
+    volumes: int
+    #: The number of external IPs consumed by the platform
+    external_ips: int
+    #: The number of CPUs consumed by the platform
+    cpus: int
```
Comment (JohnGarbutt) [inline api/azimuth/scheduling/dto.py:42]: So this is vcpus, I think. (pcpus will soon by different).

File: ui/src/components/pages/tenancy/platforms/scheduling.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+import { faCheckCircle } from '@fortawesome/free-solid-svg-icons';
+
+import { Error, Loading, formatSize, sortBy } from '../../../utils';
+
+
+// The ordering for standard quotas
+const quotaOrdering = ["machines", "volumes", "external_ips", "cpus", "ram", "storage"];
+
+
+const ProjectedQuotaProgressBar = ({ quota }) => {
```
Comment (JohnGarbutt) [inline ui/src/components/pages/tenancy/platforms/scheduling.js:19]: I think I want to review this on a dev system, if that is quick please?

File: api/azimuth/provider/openstack/provider.py [hunk 5]
```diff
@@ -443,18 +443,21 @@ def quotas(self):
...
             dto.Quota(
                 "cpus",
+                "CPUs",
                 None,
                 compute_limits.total_cores,
                 compute_limits.total_cores_used
             ),
             dto.Quota(
                 "ram",
+                "RAM",
```
Comment (JohnGarbutt) [inline api/azimuth/provider/openstack/provider.py:453]: "RAM MB"?

File: api/azimuth/provider/openstack/provider.py [hunk 6]
```diff
@@ -443,18 +443,21 @@ def quotas(self):
         quotas = [
             dto.Quota(
                 "cpus",
+                "CPUs",
```
Comment (JohnGarbutt) [inline api/azimuth/provider/openstack/provider.py:446]: Maybe vCPUs is more accurate here?

File: api/azimuth/scheduling/calculator.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+                count = parameter_values.get(count_parameter, 1)
+            else:
+                count = 1
+            if parameter.kind == "cloud.size":
+                parameter_value = self._session.find_size(parameter_value)
+                resources.add_machines(count, parameter_value)
+            elif parameter.kind == "cloud.volume_size":
+                resources.add_volumes(count, parameter_value)
+            # NOTE(mkjpryor)
+            # The cloud.volume and cloud.ip kinds represent resources that have
```
[the remark this answers is not in the document]
Comment (mkjpryor) [reply api/azimuth/scheduling/calculator.py:44]: Need to add hidden state for UI params.

File: api/azimuth/cluster_engine/dto.py [hunk 8]
```diff
@@ -51,6 +51,8 @@ class ClusterParameter:
     required: bool
     #: A default value for the parameter
     default: Any
+    #: Indicates whether or not the parameter should be hidden in the UI
+    hidden: bool
```
Comment (JohnGarbutt) [inline api/azimuth/cluster_engine/dto.py:55]: Can't this be Optional[bool]?

**Review by JohnGarbutt — APPROVED** [review]
Cool, quite a few tweaks there, but necessary ones I think.

**Pull request merged**